### DISCUSSION TITLE
Slow in getting files and dir and wondering if V2 still in development

### DISCUSSION THREAD

**Discussion by cccat6** [dokan-dev/dokany#990] [Q&A] [answered]
I currently using Dokan for my project. When the access request became more and more, it seems there is someplace that limits the performance.
CPU usage is extremely low, but the response time of getting a file or getting a directory is slow that needs a couple of seconds to get the response.
That will not happen if I simply request one file or a dir. But when my client program tries to send many requests at the same time, it will become slow.
All the data are preloaded to memory, so it is not the speed of 
I wonder why that happened and does V2 fix that? Looking for better performance on Dokan. Dokan really help my project.

**Comment by Liryna** [comment, marked as answer]
How much [threads are configured](https://github.com/dokan-dev/dokany/blob/master/dokan/dokan.h#L142) in your filesystem ?
FYI There is [this ticket](https://github.com/dokan-dev/dokany/issues/981) that would have a great impact in your situation if you are interested!

How much request are we talking about ? Can you repro with the memfs sample ?

**Reply by cccat6** [reply]
Thank you sososososo much!
I didn't there is an option of thread count. Thanks for notice that. I will try to use more threads and do some tests. I should read more documents. I just assume it will have not any limitation on threads, once a request comes in it will generate a thread to process.
I am using DokanNet in C#, I didn't actually count how many requests but I will do that and update here.
The data structure in my program is a little bit complex. There are many different kinds of cache and getting partial files from someplace else.
I should dig deeper in kernel haha
For the batching, I got some questions. How to determine the batch size and will it wait until the requests full the batch? Maybe just do a single request at first, when there are too many requests but the previous task is still in progress, then put the requests into the batch so we can process the batch next time. The batch can also be the dynamic size.
I will try to figure it out later. Hopefully I can contribute a bit.

**Reply by cccat6** [reply]
I have done the test. Threads help a lot. Previously when about 20 requests send, it will wait for more than 10 seconds since some of the requests need to get data from the database, and that cost time.
For now, each individual request will be done within 1ms, even the requests need to access the database are just a few ms.
Thanks again!

**Reply by Liryna** [reply]
@cccat6 that’s a great new! I added some doc regarding batching option here https://github.com/dokan-dev/dokany/issues/981#issuecomment-841642284 Let me know if you need anything more!
